src/model/core/fuel.jl

```diff
@@ -0,0 +1,117 @@
+
+@doc raw"""
+fuel!(EP::Model, inputs::Dict, setup::Dict)
+
+This function creates expression to account for total fuel consumption (e.g., coal, natural gas, hydrogen, etc). It also has the capability to model the piece-wise fuel consumption in part load (if data is available)
+
+***** Expressions ******
+
+The fuel consumption for power generation $vFuel_{y,t}$ is determined by power generation ($vP_{y,t}$) mutiplied by the corresponding heat rate ($Hear\_Rate_y$). 
+
+The fuel costs for power generation and start fuel for a plant $y$ at time $t$, denoted by $eCFuelOut_{y,t}$ and $eFuelStart$, is determined by fuel consumption ($vFuel_{y,t}$ and $eStartFuel$) multiplied by the fuel costs (\$/MMBTU)
+
+From above formulations, thermal generators are expected to have the same fuel consumption per generating 1 MWh electricity, regardless of minimum load or full load. However, thermal generators tend to have decreased efficiency when operating at part load, leading to higher fuel consumption per generating the same amount of electricity. To have more precise representation of fuel consumption at part load, the piecewise-linear fitting of heat input can be introduced. 
+
+```math
+\begin{aligned}
+vFuel_{y,t} >= vP_{y,t} * h_{y,x} + U_{g,t}* f_{y,x}
+\hspace{1cm} \forall y \in G, \forall t \in T, \forall x \in X
+\end{aligned}
+```
+Where $h_{y,x}$ represents incremental heat rate of a thermal generator $y$ in segment $x$ [MMBTU/MWh] and $f_{y,x}$ represents intercept of fuel consumption of a thermal generator $y$ in segment $x$ [MMBUT], and $U_{y,t}$ represents the commit status of a thermal generator $y$ at time $t$. We include at most three segments to represent the piecewise heat consumption. 
+
+Since fuel consumption has a positive value, the optimization will optimize the fuel consumption by enforcing the inequity to equal to the highest piecewise segment. When the power output is zero, the commitment variable $U_{g,t}$ will bring the intercept to be zero such that the fuel consumption is zero when thermal units are offline.
```

Comment:
I think it would be clearer to just say that only one segment is "active" at a time for any value of vP.

Reply:
I think all the segments are activated, but only one segment determines the fuel consumption at vP (i.e., the optimization will enforce fuel consumption to a segment that gives the maximum fuel usage)   
for example,
segment 1: vFuel >= 6*vP + 700;
segment 2: vFuel >= 8*vP + 100;
 
if vP = 200, vFuel in segment1 = 1900, vFuel in segment2 = 1700; then vFuel = 1900, heat rate = 9.5
if vP = 500, vFuel in segment1 = 3700, vFuel in segment2 = 4100, then vFuel = 4100, heat rate = 8.2

At the time of writing this, I decided to change "PiecewiseHeatRate" to "PiecewiseFuelUsage" as we are actually linearly approximating the fuel usage, not the heat rate.